README.md
# checkurl

[![Build Status](https://travis-ci.org/popomore/checkurl.png?branch=master)](https://travis-ci.org/popomore/checkurl) [![Coverage Status](https://coveralls.io/repos/popomore/checkurl/badge.png?branch=master)](https://coveralls.io/r/popomore/checkurl?branch=master) 

[![David Status](https://david-dm.org/popomore/checkurl.png)](https://david-dm.org/popomore/checkurl)



A tiny url status check tool

---


## Usage

### Check url

```
$ checkurl http://www.baidu.com
HEAD http://www.baidu.com/ -> 200
```

### Check combo

```
$ checkurl http://static.alipayobjects.com/static/ar/??alipay.light.base-1.4.js,alipay.security.riskMobileAccount-1.0.js
HEAD http://static.alipayobjects.com/static/ar/alipay.light.base-1.4.js -> 200
HEAD http://static.alipayobjects.com/static/ar/alipay.security.riskMobileAccount-1.0.js -> 404
```

### Check image combo

```
$ checkurl 'https://i.alipayobjects.com/combo.jpg?d=apps/24&t=10015,10016'
HEAD https://i.alipayobjects.com/common/combo/apps/24/10015.png -> 200
HEAD https://i.alipayobjects.com/common/combo/apps/24/10016.png -> 200
```

### Check file

urls.txt

```
http://static.alipayobjects.com/static/ar/alipay.light.base-1.4.js
http://static.alipayobjects.com/static/ar/alipay.security.riskMobileAccount-1.0.js
```

Check url in urls.txt

```
$ checkurl urls.txt
HEAD http://static.alipayobjects.com/static/ar/alipay.light.base-1.4.js -> 200
HEAD http://static.alipayobjects.com/static/ar/alipay.security.riskMobileAccount-1.0.js -> 404
```

### Require module

You can require `checkurl`

```
var checkurl = require('checkurl');
checkurl('url', function(data) {
  // use data
});
```

Data structure

```
[
  ['url', 200],
  ['url', 404]
]
```

if check single url, data structure is

```
['url', 200]
```

### events

1. checked

    this events will emit after every url is checked.


    ```
    checkurl('url', function(data) {
      // use data
    }).on('checked', function(data){
        
    });
    ```

2. error

    this events will emit when any url get error.

    ```
    checkurl('url', function(data) {
      // use data
    }).on('error', function(err){
        
    });
    ```

## Install

```
npm install checkurl -g
```

## License

BSD

### Thanks

https://github.com/shaoshuai0102/combocheck


HISTORY.md
# HISTORY

## 0.2.2

don't check ssl

## 0.2.1

- [#1](https://github.com/popomore/checkurl/issues/1) 解决只识别根目录 combo 的问题

## 0.2.0

- refactor
- argument support array

## 0.1.0

support image combo

## 0.0.1

First version


test/checkArray.js
var expect = require('expect.js');
var checkurl = require('..');

describe('checkArray', function() {
  it('should check array', function(done) {
    var url = [
      'https://a.alipayobjects.com/seajs/1.3.1/sea.js',
      'https://a.alipayobjects.com/seajs/1.3.1/combo.js',
      'https://a.alipayobjects.com/jquery/jquery/1.7.2/jquery.js'
    ];
    checkurl(url, function(result) {
      expect(result).to.eql([
        ['https://a.alipayobjects.com/seajs/1.3.1/sea.js', 200],
        ['https://a.alipayobjects.com/seajs/1.3.1/combo.js', 404],
        ['https://a.alipayobjects.com/jquery/jquery/1.7.2/jquery.js', 200]
      ]);
      done();
    });
  });
});

test/checkCombo.js
var expect = require('expect.js');
var checkurl = require('..');

describe('checkCombo', function() {
  it('should check assets combo', function(done) {
    var url = 'https://a.alipayobjects.com/??seajs/1.3.1/sea.js,seajs/1.3.1/combo.js,jquery/jquery/1.7.2/jquery.js';
    checkurl(url, function(result) {
      expect(result).to.eql([
        ['https://a.alipayobjects.com/seajs/1.3.1/sea.js', 200],
        ['https://a.alipayobjects.com/seajs/1.3.1/combo.js', 404],
        ['https://a.alipayobjects.com/jquery/jquery/1.7.2/jquery.js', 200]
      ]);
      done();
    });
  });

  it('should check assets combo 2', function(done) {
    var url = 'https://a.alipayobjects.com/alice/??box/1.0.0/box.css,button/1.0.0/button.css,tab/1.0.0/tab.css';
    checkurl(url, function(result) {
      expect(result).to.eql([
        ['https://a.alipayobjects.com/alice/box/1.0.0/box.css', 200],
        ['https://a.alipayobjects.com/alice/button/1.0.0/button.css', 200],
        ['https://a.alipayobjects.com/alice/tab/1.0.0/tab.css', 200]
      ]);
      done();
    });
  });

  it('should check image combo', function(done) {
    var url = 'https://i.alipayobjects.com/combo.png?d=cashier&t=CDCB,ICBC,IMCC';
    checkurl(url, function(result) {
      expect(result).to.eql([
        ['https://i.alipayobjects.com/common/combo/cashier/CDCB.png', 200],
        ['https://i.alipayobjects.com/common/combo/cashier/ICBC.png', 200],
        ['https://i.alipayobjects.com/common/combo/cashier/IMCC.png', 404]
      ]);
      done();
    });
  });

  it('should check image combo2', function(done) {
    var url = 'https://i.alipayobjects.com/combo.png?d=cashier';
    checkurl(url, function(result) {
      expect(result).to.eql([]);
      done();
    });
  });

});

test/checkSingle.js
var expect = require('expect.js');
var checkurl = require('..');

describe('checkSingle', function() {
  it('should check https', function(done) {
    var url = 'https://a.alipayobjects.com/seajs/1.2.1/sea.js';
    checkurl(url, function(result) {
      expect(result).to.eql([url, 200]);
      done();
    });
  });

  it('should check http', function(done) {
    var url = 'http://static.alipayobjects.com/seajs/1.2.1/sea.js';
    checkurl(url, function(result) {
      expect(result).to.eql([url, 200]);
      done();
    });
  });

  it('should not exist', function(done) {
    var url = 'http://static.alipayobjects.com/sea.js';
    checkurl(url, function(result) {
      expect(result).to.eql([url, 404]);
      done();
    });
  });
});

combocheck.js
var url = require('url');
var fs = require('fs');
var async = require('async');
var path = require('path');
var lineReader = require('line-reader');
var protocol = {
  http: require('http'),
  https: require('https')
};

var color = {
  C_RED: '\033[31m',
  C_BLUE: '\033[34m',
  C_GREEN: '\033[32m',
  C_RESET: '\033[0m'
};

module.exports = checkUrl;

var inputstr = process.argv[2];

if (!inputstr) exit();

if (fs.existsSync(inputstr)) {
  checkFile(file, print);
} else {
  checkUrl(inputstr, print);
}

function checkFile(file, callback) {
  lineReader.eachLine(file, function(url, last) {
    checkUrl(url, callback);
  });
}

function checkUrl(u, callback) {
  var urlObj = url.parse(u);

  // should start with http/https
  if (!urlObj.protocol) {
    var err = new Error(u + ' should start with http/https');
    callback(err);
    return;
  }

  // combo url
  if (/^\?\?/.test(urlObj.search)) {
    var prefix = u.split('??')[0];
    var arr = urlObj.search.replace(/^\?\?/, '').split(',');
    async.reduce(arr, [], function(memo, item, callback) {
      checkUrl(prefix + item, function(err, data) {
        if (err) {
          callback(err);
        }
        memo.push(data);
        callback(null, memo);
      });
    }, function(err, result) {
      callback(null, result);
    });
  } else {
    protocol[urlObj.protocol.replace(':', '')]
      .get(u, function(res) {
        callback(null, [u, res.statusCode]);
      })
      .on('error', function(e) {
        callback(e);
      });
  }
}

function exit() {
    console.error('Please input url or file');
    console.error('Example:');
    console.error('    /path/to/combocheck.js http://assets.test.alipay.net/ar/??arale.core.js,aralex.confirmbox-2.1.js');
    console.error('    /path/to/combocheck.js file.txt');
    process.exit(1);
}

function print(err, data) {
  if (err) {
    exit();
  }
  data.forEach(function(o, i) {
    var status = (o[1] == 200 ? color.C_GREEN : color.C_RED) + o[1] + color.C_RESET;
    console.log('GET ' + o[0] + ' -> ' + status);
  });
}


performance.js
var fs = require('fs');
var async = require('async');
var checkurl = require('./lib/checkurl');

fs.readFile('./data', function(err, data) {
  var arr = data.toString().split(/\r\n|\r|\n/);

  async.series([
    function(callback){
      perform(arr.slice(0, 10));
      callback(null, 'next');
    },
    function(callback){
      perform(arr.slice(11, 50));
      callback(null, 'next');
    },
    function(callback){
      perform(arr.slice(50, 200));
      callback(null, 'next');
    },
    function(callback){
      perform(arr.slice(200, 500));
      callback(null, 'next');
    },
    function(callback){
      perform(arr.slice(500, 2000));
      callback(null, 'next');
    }
  ]);

});

function perform(arr) {
  var begin = new Date().getTime();
  checkurl(arr, function(err, data) {
    var end = new Date().getTime();
    console.log(arr.length + ' times, ' + (end - begin)/arr.length + ' ms/time');
  }).on('error', function(err){
    console.log(err);
  });
}

lib/checkurl.js
var url = require('url');
var async = require('async');
var EventEmitter = require('events').EventEmitter;
var protocol = {
  http: require('http'),
  https: require('https')
};

var TEST_ASSETS_COMBO = /^https?:\/\/.*\?\?/;
var TEST_IMAGE_COMBO = /\/combo(\.png|\.jpg|32\.png)\?/;
var TEST_SINGLE_URL = /^http/;

function checkUrl(input, callback) {
  var event = new EventEmitter();

  callback || (callback = function() {});

  if (input instanceof Array) {
    checkArray(input, callback, event);

  } else if (TEST_ASSETS_COMBO.test(input)) {
    checkAssetsCombo(input, callback, event);

  } else if (TEST_IMAGE_COMBO.test(input)) {
    checkImageCombo(input, callback, event);

  } else if (TEST_SINGLE_URL.test(input)) {
    checkSingleUrl(input, callback, event);

  } else {
    callback([]);
  }
  return event;
}

module.exports =checkUrl;

// check image combo in Alipay
// checkurl('https://i.alipayobjects.com/combo.png?d=cashier&t=CDCB,ICBC,IMCC')
// return [[url1, 200], [url2, 404]]
function checkImageCombo (comboUrl, callback, event) {
  var parsed = url.parse(comboUrl);
  var query = parseQuery(parsed.query);
  var tpl = [
    parsed.protocol,
    '//',
    parsed.host,
    '/common/combo/',
    query.d,
    '/',
    '{{name}}',
    '.png'
  ].join('');

  if (query.t) {
    var arr = query.t
      .split(',').map(function(item) {
        return tpl.replace(/{{name}}/g, item);
      });
    checkArray(arr, callback, event);
  } else {
    callback([]);
  }
}

// check assets combo in Alipay
// checkurl('https://a.alipayobjects.com/??seajs/1.3.1/sea.js,seajs/1.3.1/combo.js')
// return [[url1, 200], [url2, 404]]
function checkAssetsCombo (comboUrl, callback, event) {
  var s = comboUrl.split('??');
  var prefix = s[0];
  var arr = s[1]
    .split(',')
    .map(function(item){
      return prefix + item;
    });
  checkArray(arr, callback, event);
}

// check assets of url
// checkurl([url1, url2])
// return [[url1, 200], [url2, 404]]
function checkArray (arrayUrl, callback, event) {
  async.mapLimit(arrayUrl, 80, function(singleUrl, cb) {
    if (!singleUrl) return;
    checkSingleUrl(singleUrl, function(data) {
      cb(null, data);
    }, event);
  }, function(err, results){
    callback(results);
  });
}

// check a single url
// checkurl('https://a.alipayobjects.com/seajs/1.2.1/sea.js')
// return [url, 200]
function checkSingleUrl (singleUrl, callback, event) {
  var parsed = url.parse(singleUrl);
  parsed.agent = false;
  parsed.method = 'HEAD';

  // don't check ssl
  parsed.secureOptions = require('constants').SSL_OP_NO_TLSv1_2;
  parsed.rejectUnauthorized = false;

  var request = protocol[parsed.protocol.replace(':', '')];

  // node 0.8.2 don't support string
  // https://github.com/joyent/node/blob/v0.8.2/lib/https.js#L91
  request.get(parsed, function(res) {
    var result = [singleUrl, res.statusCode];
    event.emit('checked', result);
    callback(result);
  }).on('error', function(e) {
    event.emit('error', singleUrl + " " + e);
    callback([]);
  });
}

// a=1&b=2 => {a:1, b:2}
function parseQuery(search) {
  var param = {}, a = search.split('&');
  for(var i in a) {
    var v = a[i];
    var s = v.split('=');
    param[s[0]] = s[1];
  }
  return param;
}